langchain/llms/databricks_endpoint.py

```diff
@@ -0,0 +1,224 @@
+            pass
+    assert api_token, "api_token was not set and it cannot be automatically determined."
+    return api_token
+
+class DatabricksEndpoint(LLM):
+    """LLM wrapper around a Databricks endpoint. serving endpoint or a cluster driver proxy app.
+    It supports two types of endpoints:
+    - Serving endpoint (recommended for both development and production).
+      We assume that an LLM was registered and deployed to a serving endpoint.
+      To wrap it as a LangChain LLM you must have "Can Query" permission to the endpoint.
+      Set `endpoint_name` accordingly and do not set `cluster_id` and `cluster_driver_port`.
+      The expected model signature is:
+      - inputs: `[{"name": "prompt", "type": "string"}, {"name": "stop", "type": "list[string]"}]`
+      - outputs: `[{"type": "string"}]`
+
+    - Cluster driver proxy app (recommended for rapid development).
+      One can load an LLM on a Databricks interactive cluster and start a local HTTP server
+      on the driver node to serve the model at "/" using HTTP POST with JSON input/output.
+      Please use a port number between [3000, 8000] and let the server listen to the driver IP
+      address or simply `0.0.0.0` instead of localhost only.
+      To wrap it as a LangChain LLM you must have "Can Attach To" permission to the cluster.
+      Set `cluster_id` and `cluster_driver_port` accordingly and do not set `endpoint_name`.
+      The expected server schema (using JSON schema) is:
+      - inputs: `{"type":"object","properties":{"prompt":{"type":"string"},"stop":{"type":"array","items":{"type":"string"}}},"required":["prompt"]}`
+      - outputs: `{"type":"string"}`
+
+    If the endpoint model signature is different or you want to set extra params,
+    you can use `transform_input_fn` and `transform_output_fn` to apply necessary transformations
+    before and after the query.
+    """
+
+    host: str = Field(default_factory=get_default_host)
+    """Databricks workspace hostname.
+    If not provided, the default value is determined by
+      - the ``DATABRICKS_HOST`` environment variable if present, or
+      - the hostname of the current Databricks workspace if running inside a Databricks notebook
+        attached to an interactive 
+    """
+
+    api_token: str = Field(default_factory=get_default_api_token)
+    """Databricks personal access token.
+    If not provided, the default value is determined by
+      - the ``DATABRICKS_API_TOKEN`` environment variable if present, or
+      - an automatically generated temporary token if running inside a Databricks notebook.
+    """
+
+    endpoint_name: Optional[str] = None
+    """Name of the model serving endpont.
+    You must specify the endpoint name to connect to a model serving endpoint.
+    """
+
+    cluster_id: Optional[str] = None
+    """ID of the cluster if connecting to 
+    Cannot provide both `endpoint_name` and `cluster_id`.
+    """
+
+    cluster_driver_port: Optional[str|int] = None
```

Comment:
It seems that Python 3.8 doesn't support such a union:
<img width="785" alt="image" src="https://github.com/hwchase17/langchain/assets/1097932/8d7324fd-8c13-4cfe-a1e0-1106df7d2b3f">